File: src/factories/internal/metadata/iterate_metadata_object.ts
import ts from "typescript";

import { MetadataObject } from "../../../schemas/metadata/MetadataObject";
import { MetadataObjectType } from "../../../schemas/metadata/MetadataObjectType";

import { ArrayUtil } from "../../../utils/ArrayUtil";

import { IMetadataIteratorProps } from "./IMetadataIteratorProps";
import { emplace_metadata_object } from "./emplace_metadata_object";

export const iterate_metadata_object = (
  props: IMetadataIteratorProps,
  ensure: boolean = false,
): boolean => {
  if (ensure === false) {
    const filter = (flag: ts.TypeFlags) => (props.type.getFlags() & flag) !== 0;
    if (
      !filter(ts.TypeFlags.Object) &&
      !props.type.isIntersection() &&
      (props.type as any).intrinsicName !== "object"
    )
      return false;
  }

  const obj: MetadataObjectType = emplace_metadata_object(props);
  ArrayUtil.add(
    props.metadata.objects,
    MetadataObject.create({
      type: obj,
      tags: [],
    }),
    (elem) => elem.type.name === obj.name,
  );
  return true;
};

File: packages/typia/native/core/factories/internal/metadata/iterate_metadata_object.go
package metadata

import (
  nativechecker "github.com/microsoft/typescript-go/shim/checker"
  schemametadata "github.com/samchon/typia/packages/typia/native/core/schemas/metadata"
)

func Iterate_metadata_object(props IMetadataIteratorProps, ensure ...bool) bool {
  ensured := false
  if len(ensure) != 0 {
    ensured = ensure[0]
  }
  if ensured == false {
    filter := func(flag nativechecker.TypeFlags) bool {
      return props.Type != nil && props.Type.Flags()&flag != 0
    }
    if filter(nativechecker.TypeFlagsObject) == false &&
      (props.Type == nil || props.Type.IsIntersection() == false) &&
      (props.Checker == nil || props.Type == nil || props.Checker.TypeToString(props.Type) != "object") {
      return false
    }
  }

  obj := Emplace_metadata_object(props)
  for _, elem := range props.Metadata.Objects {
    if elem.Type.Name == obj.Name {
      return true
    }
  }
  props.Metadata.Objects = append(props.Metadata.Objects, schemametadata.MetadataObject_create(schemametadata.MetadataObject{
    Type: obj,
    Tags: [][]schemametadata.IMetadataTypeTag{},
  }))
  return true
}
